### PR TITLE
SLING-9620  ResourceMapperImpl.getAllMappings does not respect multi-valued sling:alias

### PR DESCRIPTION
**Pull request by rombert**

### CODE REVIEW

**Comment by rombert** [comment]
This still needs work, there are a number of failures in the Sling ITs

```
[ERROR] Tests run: 62, Failures: 9, Errors: 0, Skipped: 0, Time elapsed: 35.074 s <<< FAILURE! - in org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest
[ERROR] testMapNamespaceMangling(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.612 s  <<< FAILURE!
org.junit.ComparisonFailure: Expect unmangled path expected:</[]sample> but was:</[virtual/]sample>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testMapNamespaceMangling(ResourceResolverGeneralTest.java:2089)

[ERROR] testResolveVirtualHostHttp80Multiple(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.581 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[http://de.host.com]/playground.html> but was:<[/de]/playground.html>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testResolveVirtualHostHttp80Multiple(ResourceResolverGeneralTest.java:716)

[ERROR] testResolveVirtualHostHttp8080(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.575 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[http://virtual.host.com:8080]/playground.html> but was:<[/virtual]/playground.html>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testResolveVirtualHostHttp8080(ResourceResolverGeneralTest.java:880)

[ERROR] testResolveVirtualHostHttps443(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.568 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[https://virtual.host.com]/playground.html> but was:<[/virtual]/playground.html>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testResolveVirtualHostHttps443(ResourceResolverGeneralTest.java:950)

[ERROR] testResolveVirtualHostHttps4443(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.563 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[https://virtual.host.com:4443]/playground.html> but was:<[/virtual]/playground.html>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testResolveVirtualHostHttps4443(ResourceResolverGeneralTest.java:984)

[ERROR] testMapURLEscaping(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.581 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[http://internal.host.com]/sample%20with%20spa...> but was:<[/internal]/sample%20with%20spa...>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testMapURLEscaping(ResourceResolverGeneralTest.java:2029)

[ERROR] testResolveResourceInternalRedirectExact(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.55 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[https://localhost/virtua]l> but was:<[/virtual.htm]l>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testResolveResourceInternalRedirectExact(ResourceResolverGeneralTest.java:600)

[ERROR] testResolveVirtualHostHttp80(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.561 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[http://virtual.host.com]/playground.html> but was:<[/virtual]/playground.html>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testResolveVirtualHostHttp80(ResourceResolverGeneralTest.java:667)

[ERROR] testResolveVirtualHostHttpVsHttps(org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest)  Time elapsed: 0.555 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<[http://www.host.com]/playground.html> but was:<[/page]/playground.html>
	at org.apache.sling.launchpad.webapp.integrationtest.resourceresolver.ResourceResolverGeneralTest.testResolveVirtualHostHttpVsHttps(ResourceResolverGeneralTest.java:1043)

```

File: src/test/java/org/apache/sling/resourceresolver/impl/mapping/PathGeneratorTest.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    public void subPathWithMultipleAliases() {
+        
+        PathGenerator builder = new PathGenerator();
+        builder.insertSegment(emptyList(), "bar");
+        builder.insertSegment(asList("alias1", "alias2"), "foo");
+        
+        List<String> paths = builder.generatePaths();
+        
+        assertThat(paths, Matchers.hasSize(2));
+        assertThat(paths, Matchers.hasItems("/alias1/bar", "/alias2/bar"));
```
Comment (jsedding) [inline src/test/java/org/apache/sling/resourceresolver/impl/mapping/PathGeneratorTest.java:90]: @rombert What is the reason that `PathGenerator` does not generate permutations of the path including the node-name? I.e. I'd expect `Matchers.hasItems("/alias1/bar", "/alias2/bar", "/foo/bar")` (probably in this order, if the first entry is what should happen when mapping a resource path to a URL path).

Comment (rombert) [reply src/test/java/org/apache/sling/resourceresolver/impl/mapping/PathGeneratorTest.java:90]: The `PathGenerator` is a bit special-cased as it generates the path-based variant only if no other aliases exist. Would it be clearer it it only generated the alias-based combinations, as the node name only variant is already handled differently in the `ResourceMapperImpl`?

**Comment by rombert** [comment]
@anchela , @jsedding - this is ready for review, I added tests for optimise alias resolution enabled and disabled; Sling ITs pass.

It would be great if could review the changes.

**Comment by rombert** [comment]
I've now updated the changes to cover SLING-9623 as well.

**Comment by rombert** [comment]
Hm, a failing test slipped by. I will look into it.

**Comment by rombert** [comment]
Testing should be ok now.

File: src/test/java/org/apache/sling/resourceresolver/impl/mapping/PathGeneratorTest.java (hunk 1 above)
Comment (rombert) [reply src/test/java/org/apache/sling/resourceresolver/impl/mapping/PathGeneratorTest.java]: FWIW, the behaviour is now the one you suggested, due to the special case of decorated resources. We now generate all combinations in this class and then remove extra ones if needed.

Comment (jsedding) [reply src/test/java/org/apache/sling/resourceresolver/impl/mapping/PathGeneratorTest.java]: Thanks @rombert for the heads-up. I believe that makes the code slightly easier to understand (principle of least surprise).

**Pull request merged**